File: Back End/TravelPlanner.API/Services/Auth/AuthService.cs
using System.Security.Cryptography;
using System.Text;
using Microsoft.EntityFrameworkCore;
using TravelPlanner.API.Data;
using TravelPlanner.API.DTOs.Auth;
using TravelPlanner.API.Models;

namespace TravelPlanner.API.Services.Auth;

public class AuthService : IAuthService
{
    private readonly TravelPlannerDbContext _dbContext;
    private readonly IJwtTokenService _jwtTokenService;
    private readonly IConfiguration _configuration;

    public AuthService(TravelPlannerDbContext dbContext, IJwtTokenService jwtTokenService, IConfiguration configuration)
    {
        _dbContext = dbContext;
        _jwtTokenService = jwtTokenService;
        _configuration = configuration;
    }

    public async Task<AuthResponseDto> RegisterAsync(RegisterDto dto)
    {
        if (dto.Password != dto.ConfirmPassword)
        {
            throw new ArgumentException("Mật khẩu xác nhận không khớp.");
        }

        var existingUser = await _dbContext.Users.FirstOrDefaultAsync(u => u.Email == dto.Email);
        if (existingUser != null)
        {
            throw new ArgumentException("Email này đã được sử dụng.");
        }

        if (!string.IsNullOrEmpty(dto.Phone))
        {
            var existingPhone = await _dbContext.Users.FirstOrDefaultAsync(u => u.Phone == dto.Phone);
            if (existingPhone != null)
            {
                throw new ArgumentException("Số điện thoại này đã được sử dụng.");
            }
        }

        var user = new User
        {
            FullName = dto.FullName,
            Email = dto.Email,
            Phone = dto.Phone,
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(dto.Password),
            IsActive = true
        };

        _dbContext.Users.Add(user);
        await _dbContext.SaveChangesAsync();

        return await CreateAuthResponseAsync(user);
    }

    public async Task<AuthResponseDto> LoginAsync(LoginDto dto)
    {
        var user = await _dbContext.Users.FirstOrDefaultAsync(u => u.Email == dto.Email);
        if (user == null || !BCrypt.Net.BCrypt.Verify(dto.Password, user.PasswordHash))
        {
            throw new UnauthorizedAccessException("Email hoặc mật khẩu không chính xác.");
        }

        if (!user.IsActive)
        {
            throw new UnauthorizedAccessException("Tài khoản của bạn đã bị vô hiệu hóa.");
        }

        return await CreateAuthResponseAsync(user);
    }

    public async Task<AuthResponseDto> RefreshTokenAsync(RefreshTokenDto dto)
    {
        var tokenHash = HashToken(dto.RefreshToken);

        var savedRefreshToken = await _dbContext.RefreshTokens
            .Include(rt => rt.User)
            .FirstOrDefaultAsync(rt => rt.TokenHash == tokenHash);

        if (savedRefreshToken == null || savedRefreshToken.RevokedAt != null || savedRefreshToken.ExpiresAt <= DateTime.UtcNow)
        {
            throw new UnauthorizedAccessException("Refresh token không hợp lệ hoặc đã hết hạn.");
        }

        // Cập nhật token hiện tại thành đã sử dụng (revoked)
        savedRefreshToken.RevokedAt = DateTime.UtcNow;

        return await CreateAuthResponseAsync(savedRefreshToken.User);
    }

    public async Task LogoutAsync(string token)
    {
        var tokenHash = HashToken(token);
        var savedRefreshToken = await _dbContext.RefreshTokens.FirstOrDefaultAsync(rt => rt.TokenHash == tokenHash);
        
        if (savedRefreshToken != null && savedRefreshToken.RevokedAt == null)
        {
            savedRefreshToken.RevokedAt = DateTime.UtcNow;
            await _dbContext.SaveChangesAsync();
        }
    }

    public async Task<UserProfileDto?> GetUserProfileAsync(ulong userId)
    {
        var user = await _dbContext.Users.FindAsync(userId);
        if (user == null) return null;

        return new UserProfileDto
        {
            Id = user.Id,
            FullName = user.FullName,
            Email = user.Email,
            Phone = user.Phone,
            AvatarUrl = user.AvatarUrl,
            Role = user.Role
        };
    }

    private async Task<AuthResponseDto> CreateAuthResponseAsync(User user)
    {
        var accessToken = _jwtTokenService.GenerateAccessToken(user);
        var refreshToken = _jwtTokenService.GenerateRefreshToken();
        var refreshTokenHash = HashToken(refreshToken);

        var expiryDays = int.Parse(_configuration["JwtSettings:RefreshTokenExpirationDays"] ?? "7");

        var newRefreshToken = new RefreshToken
        {
            UserId = user.Id,
            TokenHash = refreshTokenHash,
            ExpiresAt = DateTime.UtcNow.AddDays(expiryDays)
        };

        _dbContext.RefreshTokens.Add(newRefreshToken);
        await _dbContext.SaveChangesAsync();

        return new AuthResponseDto
        {
            Token = accessToken,
            RefreshToken = refreshToken,
            User = new UserProfileDto
            {
                Id = user.Id,
                FullName = user.FullName,
                Email = user.Email,
                Phone = user.Phone,
                AvatarUrl = user.AvatarUrl,
                Role = user.Role
            }
        };
    }

    private string HashToken(string token)
    {
        using var sha256 = SHA256.Create();
        var bytes = Encoding.UTF8.GetBytes(token);
        var hash = sha256.ComputeHash(bytes);
        return Convert.ToBase64String(hash);
    }
}

File: backend/src/services/authService.js
import bcrypt from "bcryptjs";
import prisma from "../config/db.js";
import {
    generateAccessToken,
    generateRefreshToken,
    hashToken
} from "./jwtService.js";

/**
 * Đăng ký user mới
 * Tương đương AuthService.RegisterAsync trong C#
 */
export async function register(dto) {
    const { fullName, email, phone, password, confirmPassword } = dto;

    if (password !== confirmPassword) {
        throw { status: 400, message: "Mật khẩu xác nhận không khớp." };
    }

    const existingEmail = await prisma.users.findUnique({ where: { email } });
    if (existingEmail) {
        throw { status: 400, message: "Email này đã được sử dụng." };
    }

    if (phone) {
        const existingPhone = await prisma.users.findUnique({ where: { phone } });
        if (existingPhone) {
            throw { status: 400, message: "Số điện thoại này đã được sử dụng." };
        }
    }

    const passwordHash = await bcrypt.hash(password, 10);

    const user = await prisma.users.create({
        data: {
            full_name: fullName,
            email,
            phone: phone || null,
            password_hash: passwordHash,
            role: "user",
            is_active: true
        }
    });

    return await createAuthResponse(user);
}

/**
 * Đăng nhập
 * Tương đương AuthService.LoginAsync trong C#
 */
export async function login(dto) {
    const { email, password } = dto;

    const user = await prisma.users.findUnique({ where: { email } });
    if (!user) {
        throw { status: 401, message: "Email hoặc mật khẩu không chính xác." };
    }

    const isValid = await bcrypt.compare(password, user.password_hash);
    if (!isValid) {
        throw { status: 401, message: "Email hoặc mật khẩu không chính xác." };
    }

    if (!user.is_active) {
        throw { status: 401, message: "Tài khoản của bạn đã bị vô hiệu hóa." };
    }

    return await createAuthResponse(user);
}

/**
 * Refresh token
 * Tương đương AuthService.RefreshTokenAsync trong C#
 */
export async function refreshToken(refreshTokenValue) {
    const tokenHash = hashToken(refreshTokenValue);

    const savedToken = await prisma.refresh_tokens.findUnique({
        where: { token_hash: tokenHash },
        include: { users: true }
    });

    if (
        !savedToken ||
        savedToken.revoked_at !== null ||
        savedToken.expires_at <= new Date()
    ) {
        throw { status: 401, message: "Refresh token không hợp lệ hoặc đã hết hạn." };
    }

    // Revoke token hiện tại
    await prisma.refresh_tokens.update({
        where: { id: savedToken.id },
        data: { revoked_at: new Date() }
    });

    return await createAuthResponse(savedToken.users);
}

/**
 * Đăng xuất - revoke refresh token
 * Tương đương AuthService.LogoutAsync trong C#
 */
export async function logout(refreshTokenValue) {
    const tokenHash = hashToken(refreshTokenValue);

    const savedToken = await prisma.refresh_tokens.findUnique({
        where: { token_hash: tokenHash }
    });

    if (savedToken && savedToken.revoked_at === null) {
        await prisma.refresh_tokens.update({
            where: { id: savedToken.id },
            data: { revoked_at: new Date() }
        });
    }
}

/**
 * Lấy profile user
 * Tương đương AuthService.GetUserProfileAsync trong C#
 */
export async function getUserProfile(userId) {
    const user = await prisma.users.findUnique({
        where: { id: BigInt(userId) }
    });

    if (!user) return null;

    return buildUserProfile(user);
}

// ===== HELPERS =====

async function createAuthResponse(user) {
    const accessToken = generateAccessToken(user);
    const newRefreshToken = generateRefreshToken();
    const refreshTokenHash = hashToken(newRefreshToken);

    const expiryDays = parseInt(process.env.REFRESH_TOKEN_EXPIRY_DAYS || "7", 10);

    await prisma.refresh_tokens.create({
        data: {
            user_id: user.id,
            token_hash: refreshTokenHash,
            expires_at: new Date(Date.now() + expiryDays * 24 * 60 * 60 * 1000)
        }
    });

    return {
        token: accessToken,
        refreshToken: newRefreshToken,
        user: buildUserProfile(user)
    };
}

function buildUserProfile(user) {
    return {
        id: Number(user.id),
        fullName: user.full_name,
        email: user.email,
        phone: user.phone || null,
        avatarUrl: user.avatar_url || null,
        role: user.role === "admin" ? 1 : 0
    };
}
